### PR TITLE
Replace "Linked Agent By Role" - new Plugin for filtering Typed Relations in search API

### PR DESCRIPTION
**Pull request by rosiel**
* REPLACES #88* (my apologies for git silliness)

* Other Relevant Links:
* discussed recently with @seth-shaw-asu and Chris Day (no @?)

# What does this Pull Request do?

Creates new "fields" (in search api) for your Typed Relation fields. You can filter what relation types get included.  

This does the same thing as the [Linked Agent By Role](https://github.com/Islandora-Devops/islandora_install_profile_demo/blob/main/modules/islandora_install_profile_demo_core/src/Plugin/search_api/processor/LinkedAgentByType.php) plugin on the Islandora Install Profile Demo, except without the hard-coding and without cluttering the Search API field with 271 different fields.

# What's new?

New Search API Processor plugin!

* Creates a new field (under "general" - FIXME!!) for each configured Typed Relation field. For example there is one for "Linked Agent [islandora_object]"
* When you select this field, it lets you choose one or more "Relation Types" from the list that is currently configured for that field.

Selecting the new field

Configuring the new field

* Does this change require documentation to be updated?  Sure!
* Does this change add any new dependencies? no!
* Does this change require any other modifications to be made to the repository
 (i.e. Regeneration activity, etc.)? nah.
* Could this change impact execution of existing code? no.

# How should this be tested?

A description of what steps someone could take to:
* Reproduce the problem you are fixing (if applicable)
* Test that the Pull Request does what is intended.
* Please be as detailed as possible.
* Good testing instructions help get your PR completed faster.

# Additional Notes:
Any additional information that you think would be helpful when reviewing this
 PR.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @Islandora/committers

### CODE REVIEW

**Comment by rosiel** [comment]
Can be tested, feedback welcome. 

I'm going to try to move these fields out of search api's "General" section, and into the individual sections for "Content", "Media", etc.

**Comment by rosiel** [comment]
With the last commit, the fields no longer show in "General" but in the appropriate "datasource" (i.e. Content, Taxonomy Term, etc... they show up with their partner fields.)

If you already had started testing before this commit, you need to remove existing fields of this type, and clear the index and "rebuild tracking information". I had trouble because it had cached the old (invalid) field definition.

Screen Shot 2022-11-23 at 10 00 22 PM

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  public function getPropertyDefinitions(DatasourceInterface $datasource = NULL): array {
+    $properties = [];
+
+    if (!$datasource || !$datasource->getEntityTypeId()) {
+      return $properties;
+    }
+
+    $entity_type = $datasource->getEntityTypeId();
+    // Get all configured typed relation fields.
+    $fields = \Drupal::entityTypeManager()->getStorage('field_config')->loadByProperties([
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: DI as you pointed out in Slack.

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+          $search_api_fields = $this->getFieldsHelper()
+            ->filterForPropertyPath($search_api_fields, $item->getDatasourceId(), $property_path);
+          foreach ($search_api_fields as $search_api_field) {
+            // Load entity's field values.
+            $vals = $content_entity->$field_name->getValue();
+            foreach ($vals as $element) {
+              $rel_type = $element['rel_type'];
+              if (in_array($rel_type, $search_api_field->getConfiguration()['rel_types'])) {
+                $tid = $element['target_id'];
+                $taxo_term = \Drupal::entityTypeManager()
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: DI.

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+ * Adds filterable fields for each Typed Relation field.
+ *
+ * @SearchApiProcessor(
+ *   id = "typed_relation_filtered",
+ *   label = @Translation("Typed Relation, filtered by type"),
+ *   description = @Translation("Filter Typed Relation fields by type"),
+ *   stages = {
+ *     "add_properties" = 0,
+ *   },
+ *   locked = true,
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: As discussed in Slack this should probably be `false`.

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    if (!$datasource || !$datasource->getEntityTypeId()) {
+      return $properties;
+    }
+
+    $entity_type = $datasource->getEntityTypeId();
+    // Get all configured typed relation fields.
+    $fields = \Drupal::entityTypeManager()->getStorage('field_config')->loadByProperties([
+      'entity_type' => $entity_type,
+      'field_type' => 'typed_relation',
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php:87]: Would it be possible to store the field name in here so the logic when it comes around to be used is simple/less expensive?

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      if ($field->getType() == 'typed_relation') {
+        $field_name = $field->getName();
+        if (!$content_entity->get($field_name)->isEmpty()) {
+          // See if this field is being indexed.
+          $property_path = 'typed_relation_filter__' . str_replace('.', '__', $field->id());
+          $search_api_fields = $this->getFieldsHelper()
+            ->filterForPropertyPath($search_api_fields, $item->getDatasourceId(), $property_path);
+          foreach ($search_api_fields as $search_api_field) {
+            // Load entity's field values.
+            $vals = $content_entity->$field_name->getValue();
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: Should be able to use `referencedEntities` here without the need to load up things I think.

Comment (rosiel) [reply src/Plugin/search_api/processor/TypedRelationFiltered.php]: I need the `rel_type`, i'm not sure that would be in the referenced entities list.

**Comment by simonhm** [comment]
@rosiel 
We're testing this patch, and unfortunately this's not working as intended. 
After patching this PR, new search API processor is showed up. I enabled it and was able to add new field Linked Agent (filtered by type) as same as your instruction.
Then when I re-index, I got the below messages in Recent Logs:

`Notice: Undefined index: bundle in Drupal\controlled_access_terms\Plugin\search_api\processor\TypedRelationFiltered->addFieldValues() (line 137 of /var/www/html/drupal/web/modules/contrib/controlled_access_terms/src/Plugin/search_api/processor/TypedRelationFiltered.php)
#0 /var/www/html/drupal/web/core/includes/bootstrap.inc(347): _drupal_error_handler_real(8, 'Undefined index...', '/var/www/html/d...', 137)
#1 /var/www/html/drupal/web/modules/contrib/controlled_access_terms/src/Plugin/search_api/processor/TypedRelationFiltered.php(137): _drupal_error_handler(8, 'Undefined index...', '/var/www/html/d...', 137, Array) ...`

And the facet for this new field is not displaying/working as intended after I do a search.
Simon.

**Comment by rosiel** [comment]
Thank you so much! I think I've fixed that now.

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    foreach ($relevant_search_api_fields as $search_api_field) {
+      $field_config = $search_api_field->getConfiguration();
+
+      // Exit if we're on the wrong bundle or the field isn't set.
+      if (($content_entity->bundle() != $field_config['bundle'])
+        || !$content_entity->hasField($field_config['base_field'])) {
+        return;
+      }
+
+      $vals = $content_entity->$field_config['base_field'];
```
Comment (jordandukart) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: This is always empty in its current form.

Should it be 
```suggestion
      $vals = $content_entity->get($field_config['base_field'])->getValue();
```

File: src/Plugin/search_api/processor/TypedRelationFiltered.php [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    foreach ($relevant_search_api_fields as $search_api_field) {
+      $field_config = $search_api_field->getConfiguration();
+
+      // Exit if we're on the wrong bundle or the field isn't set.
+      if (($content_entity->bundle() != $field_config['bundle'])
+        || !$content_entity->hasField($field_config['base_field'])) {
+        return;
+      }
+
+      $vals = $content_entity->get($field_config['base_field'])->getValue;
```
Comment (simonhm) [inline src/Plugin/search_api/processor/TypedRelationFiltered.php]: @rosiel 
This line should be: 
`$vals = $content_entity->get($field_config['base_field'])->getValue();`
You're missing the () in getValue()

Within (), this patch is working for us like a charm. Thanks!

Comment (rosiel) [reply src/Plugin/search_api/processor/TypedRelationFiltered.php]: Thank you!

File: src/Plugin/search_api/processor/TypedRelationFiltered.php (hunk 6 above)
Comment (rosiel) [reply src/Plugin/search_api/processor/TypedRelationFiltered.php]: Thanks! I have tested, and it is now working for me (with @simonhm's correction - whoops!)

**Comment by jordandukart** [comment]
Configuration schema changes here are the same as in https://github.com/Islandora/islandora/pull/913#discussion_r1089461386. I think we may be able to re-work how these are constructed. I may give this a go and update this further.

**Comment by seth-shaw-asu** [comment]
@jordandukart, were you able to re-work the configuration schema changes as mentioned in https://github.com/Islandora/controlled_access_terms/pull/89#issuecomment-1426021950, or should we move forward with testing and merging?

**Comment by wgilling** [comment]
I will add "review this" PR to my to-do list for this week sometime.

**Comment by wgilling** [comment]
I must have something in my local "bd-isle" project that is interfering with seeing these. Would I be able to see the "Linked Agent (filtered by type) [islandora] typed_relation_filter__node__islandora_object_field_linked_agent" on the Solr index | Fields page https://islandora.traefik.me/admin/config/search/search-api/index/default_solr_index/fields?

image

**Comment by rosiel** [comment]
You might have to enable it first, under the "Processors" tab.

**Comment by wgilling** [comment]
Linked Agent by Type "adds the item's linked agent separately by type" was checked. This is probably conflicting with something in the bd-isle flavor site I pulled this feature branch of controlled access terms into. How do I spin up a vanilla islandora site?

**Review by ajstanley — APPROVED** [review]
Once the processor has been added, the new form shows up where it should.  Tested with Contributor, and an arbitrary new field.

**Pull request merged**